pkg/lib/harness/llm-harness.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2024 The KitOps Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package harness
+
+import (
+	"fmt"
+	"kitops/pkg/lib/constants"
+	"kitops/pkg/output"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"syscall"
+)
+
+type LLMHarness struct {
+	Host       string
+	Port       int
+	ConfigHome string
+}
+
+func (harness *LLMHarness) Init() {
+	extractLibraries(constants.HarnessPath(harness.ConfigHome), "llama.cpp/build/*/*/*/bin/**")
+}
+
+func (harness *LLMHarness) Start(modelPath string) error {
+	pidFile := filepath.Join(constants.HarnessPath(harness.ConfigHome), constants.HarnessProcessFile)
+
+	if _, err := os.Stat(pidFile); !os.IsNotExist(err) {
+		// Attempt to read the PID from the file.
+		pid, err := readPIDFromFile(pidFile)
+		if err != nil {
+			return fmt.Errorf("failed to read PID file: %w", err)
+		}
+		// Check if the process is still running.
+		if isProcessRunning(pid) {
+			return fmt.Errorf("A server process with PID %d is already running", pid)
```

Comment:
Is there any way to instead return information about the running server? E.g.

```
Development server is already running at http://127.0.0.1:51021
Use "kit dev --stop" to stop the development server
```

Reply:
We could save the URL to somewhere I suppose.